src/validr/_validator.pyx

```diff
@@ -129,14 +130,21 @@ def list_validator(compiler, items=None, int minlen=0, int maxlen=1024,
         except TypeError:
             raise Invalid('not list')
         result = []
+        keys = set()
         cdef int i = -1
         for i, x in value:
             if i >= maxlen:
                 raise Invalid('list length must <= %d' % maxlen)
             with mark_index(i):
                 v = inner(x) if inner is not None else copy.deepcopy(x)
-                if unique and v in result:
-                    raise Invalid('not unique')
+                if unique:
+                    if unique_key_helper:
+                        key = compiler.helpers.get(unique_key_helper)(v)
+                        if key in keys:
+                            raise Invalid('not unique')
+                        keys.add(key)
+                    elif v in result:
```

Comment:
When v is a scalar value and no unique_key_helper, `v in result` is still an O(N) operation?

Reply:
Yeah, that's left unmodified.

Reply:
we can know if v is a scalar value from the inner schema, and speed up the unique check in this case.

Reply:
Best of all may be if we can guess the key function based on the inner schema. I don't know the full way right now but my guess is it's relatively simple, based on that we go from down to top. Like define hashing function for all validator types (e.g. for dicts/lists there we have hashes of their inner values the function looks trivial too.) That, from my guess, will speed up unique checks VERY magnificiently and be an ALMOST out-of-the-box solution. If the user specifies hashing for the inner validators, it becomes out-of-the-box again. AND it also becomes raliable and fast.

Reply:
But, of course, it also takes more work to make it working now. But the task itself looks interesting enough. Not sure i'll be able to implement it myself, as i'm a little busy at work with the task there i use your validator in production code, but it's just a small part and i'm not sure i'll be able to spend much time in helping to make your code even better (And for now it's good enough).
Anyway, thanks for your job done. This is a great piece of code indeed.
By the way. You're writing in python only? Maybe you'll be interested in trying Nim? It's syntax is much like python's put the language itself is much more low-level (and FAST!), and yet comfortable, flexible and elegant.
If you're up to, that's nim home site: https://nim-lang.org